core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3143,63 +3143,35 @@ private void mountInternal(RpcContext rpcContext, LockedInodePath inodePath, All
-    mountInternal(rpcContext, inodePath, ufsPath, mountId, context);
+    // get UfsManager prepared
+    mUfsManager.addMount(mountId, new AlluxioURI(ufsPath.toString()),
+        new UnderFileSystemConfiguration(
+            Configuration.global(), context.getOptions().getReadOnly())
+            .createMountSpecificConf(context.getOptions().getPropertiesMap()));
+    prepareForMount(ufsPath, mountId, context);
+
     boolean loadMetadataSucceeded = false;
     try {
       // This will create the directory at alluxioPath
-      InodeSyncStream.loadDirectoryMetadata(rpcContext,
+      InodeSyncStream.loadMountPointDirectoryMetadata(rpcContext,
           inodePath,
           LoadMetadataContext.mergeFrom(
               LoadMetadataPOptions.newBuilder().setCreateAncestors(false)),
-          mMountTable,
+          context.getOptions().getShared(), ufsPath, mUfsManager.get(mountId),
           this);
       loadMetadataSucceeded = true;
     } finally {
-      if (!loadMetadataSucceeded) {
-        mMountTable.delete(rpcContext, inodePath.getUri(), true);
-      }
-    }
-  }
-
-  /**
-   * Updates the mount table with the specified mount point. The mount options may be updated during
-   * this method.
-   *
-   * @param journalContext the journal context
-   * @param inodePath the Alluxio mount point
-   * @param ufsPath the UFS endpoint to mount
-   * @param mountId the mount id
-   * @param context the mount context (may be updated)
-   */
-  private void mountInternal(Supplier<JournalContext> journalContext, LockedInodePath inodePath,
-      AlluxioURI ufsPath, long mountId, MountContext context)
-      throws FileAlreadyExistsException, InvalidPathException, IOException {
-    AlluxioURI alluxioPath = inodePath.getUri();
-    // Adding the mount point will not create the UFS instance and thus not connect to UFS
-    mUfsManager.addMount(mountId, new AlluxioURI(ufsPath.toString()),
-        new UnderFileSystemConfiguration(
-            Configuration.global(), context.getOptions().getReadOnly())
-            .createMountSpecificConf(context.getOptions().getPropertiesMap()));
-    try {
-      prepareForMount(ufsPath, mountId, context);
-      // Check that the alluxioPath we're creating doesn't shadow a path in the parent UFS
-      MountTable.Resolution resolution = mMountTable.resolve(alluxioPath);
-      try (CloseableResource<UnderFileSystem> ufsResource =
-          resolution.acquireUfsResource()) {
-        String ufsResolvedPath = resolution.getUri().getPath();
-        if (ufsResource.get().exists(ufsResolvedPath)) {
-          throw new IOException(MessageFormat.format(
-              "Mount path {0} shadows an existing path {1} in the parent underlying filesystem",
-              alluxioPath, ufsResolvedPath));
-        }
+      if (loadMetadataSucceeded) {
+        // As we have verified the mount operation by calling MountTable.verifyMount, there won't
+        // be any error thrown when doing MountTable.add
```

Comment:
Im wondering if there could be a concurrency issue here with the the verify mount? I.e. someone could mount the same UFS path concurrently after verify is called, then you could have 2 mounts of the same UFS path in different directories in Alluxio?

Reply:
You are right. I will design a synchronized mechanism to solve this problem.